crates/native/src/media/mod.rs

```diff
@@ -0,0 +1,646 @@
+        &mut self,
+        caps: &api::AudioConstraints,
+    ) -> anyhow::Result<Arc<AudioSource>> {
+        let device_id = if let Some(device_id) = caps.device_id.clone() {
+            device_id.into()
+        } else {
+            // `AudioDeviceModule` is not capturing anything at the moment,
+            // so we will use first available device (with `0` index).
+            if self.audio_device_module.recording_devices() < 1 {
+                bail!("Cannot find any available audio input device");
+            }
+
+            self.audio_device_module.recording_device_name(0)?.1.into()
+        };
+
+        let Some(device_index) =
+            self.get_index_of_audio_recording_device(&device_id)?
+        else {
+            bail!(
+                "Cannot find audio device with the specified ID `{device_id}`",
+            );
+        };
+
+        let src = if let Some(src) = self.audio_sources.get(&device_id) {
+            src.update_audio_processing(&caps.processing);
+            Arc::clone(src)
+        } else {
+            let processing =
+                sys::AudioProcessing::new((&caps.processing).into())?;
+
+            let src = Arc::new(AudioSource::new(
+                device_id.clone(),
+                self.audio_device_module
+                    .create_audio_source(device_index, &processing)?,
+                processing,
+            ));
+            self.audio_sources.insert(device_id, Arc::clone(&src));
+
+            src
+        };
+
+        Ok(src)
+    }
+
+    /// Returns the [readyState][0] property of the media track by its ID and
+    /// media type.
+    ///
+    /// [0]: https://w3.org/TR/mediacapture-streams#dfn-readystate
+    #[must_use]
+    pub fn track_state(
+        &self,
+        id: String,
+        track_origin: TrackOrigin,
+        kind: api::MediaType,
+    ) -> api::TrackState {
+        match kind {
+            api::MediaType::Audio => {
+                let id = AudioTrackId::from(id);
+                self.audio_tracks
+                    .get(&(id, track_origin))
+                    .map_or_else(|| api::TrackState::Ended, |t| t.state())
+            }
+            api::MediaType::Video => {
+                let id = VideoTrackId::from(id);
+                self.video_tracks
+                    .get(&(id, track_origin))
+                    .map_or_else(|| api::TrackState::Ended, |t| t.state())
+            }
+        }
+    }
+
+    /// Returns the [width] property of the media track by its ID and origin.
+    ///
+    /// Blocks until the [width] is initialized.
+    ///
+    /// [width]: https://w3.org/TR/mediacapture-streams#dfn-width
+    #[must_use]
+    pub fn track_width(
+        &self,
+        id: String,
+        track_origin: TrackOrigin,
+    ) -> Option<i32> {
+        let id = VideoTrackId::from(id);
+
+        self.video_tracks
+            .get(&(id, track_origin))
+            .map(|t| t.kind.dimensions.wait().read().unwrap().width())
```

Comment:
Another reason to not access things like that with public fields. In every place where we're getting access to `dimensions` we need to write this ugly `.wait().read().unwrap()`. Also by doing that, we're giving uncontrolled access to mutation of interior fields. Better to hide those details of implementations under the hood, so we can change underlying primitives and impls easily (e.g. change `RwLock` to `Mutex`).